lib/cuda/cudaFRBBeamformer.cpp

```diff
@@ -1,280 +1,222 @@
+
+private:
+    // Julia's `CuDevArray` type
+    template<typename T, std::int64_t N>
+    struct CuDeviceArray {
+        T* ptr;
+        std::int64_t maxsize; // bytes
+        std::int64_t dims[N]; // elements
+        std::int64_t len;     // elements
+        CuDeviceArray(void* const ptr, const std::size_t bytes) :
+            ptr(static_cast<T*>(ptr)), maxsize(bytes), dims{std::int64_t(maxsize / sizeof(T))},
+            len(maxsize / sizeof(T)) {}
+    };
+    using kernel_arg = CuDeviceArray<int32_t, 1>;
+
+    // Kernel design parameters:
+    static constexpr int cuda_beam_layout_M = 48;
+    static constexpr int cuda_beam_layout_N = 48;
+    static constexpr int cuda_dish_layout_M = 24;
+    static constexpr int cuda_dish_layout_N = 24;
+    static constexpr int cuda_downsampling_factor = 40;
+    static constexpr int cuda_number_of_complex_components = 2;
+    static constexpr int cuda_number_of_dishes = 512;
+    static constexpr int cuda_number_of_frequencies = 256;
+    static constexpr int cuda_number_of_polarizations = 2;
+    static constexpr int cuda_number_of_timesamples = 2064;
+
+    // Kernel compile parameters:
+    static constexpr int minthreads = 768;
+    static constexpr int blocks_per_sm = 1;
+
+    // Kernel call parameters:
+    static constexpr int threads_x = 32;
+    static constexpr int threads_y = 24;
+    static constexpr int blocks = 256;
+    static constexpr int shmem_bytes = 76896;
+
+    // Kernel name:
+    const char* const kernel_symbol =
+        "_Z3frb13CuDeviceArrayI7Int16x2Li1ELi1EES_I9Float16x2Li1ELi1EES_I6Int4x8Li1ELi1EES_IS1_"
+        "Li1ELi1EES_I5Int32Li1ELi1EE";
+
+    // Kernel arguments:
+    static constexpr std::size_t S_length = 2304UL;
+    static constexpr std::size_t W_length = 1179648UL;
+    static constexpr std::size_t E_length = 541065216UL;
+    static constexpr std::size_t I_length = 239616UL;
+    static constexpr std::size_t info_length = 786432UL;
+
+    // Runtime parameters:
+
+    // GPU memory:
+    const std::string S_memname;
+    const std::string W_memname;
+    const std::string E_memname;
+    const std::string I_memname;
+    const std::string info_memname;
+
+    // Host-side buffer arrays
+    std::vector<std::vector<std::int32_t>> host_info;
+
+    // Declare extra variables (if any)
+};
 
-static const size_t sizeof_float16_t = 2;
+REGISTER_CUDA_COMMAND(cudaFRBBeamformer);
 
 cudaFRBBeamformer::cudaFRBBeamformer(Config& config, const std::string& unique_name,
                                      bufferContainer& host_buffers, cudaDeviceInterface& device) :
-    cudaCommand(config, unique_name, host_buffers, device, "FRB_beamformer", "frb.ptx") {
-
-    // HACK -- at the very beginning of the run, we pretend we have 40 samples of padding,
-    // so that we get 51 outputs (otherwise we would get 50), so that the rechunker produces
-    // an output frame of 256 outputs after every 5 input frames.
-    padded_samples = 40;
```

Comment:
I have just taken all of Erik's autogenerated code here, which means we've dropped the padding/overlap handling, but that will get pushed upstream to Erik's IndexSpaces package where this C++ code is auto-generated.